File: teamapps-client/projector-combobox/src/main/java/org/teamapps/projector/components/trivial/multicurrency/CurrencyField.java
/*-
 * ========================LICENSE_START=================================
 * TeamApps
 * ---
 * Copyright (C) 2014 - 2022 TeamApps.org
 * ---
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * =========================LICENSE_END==================================
 */
package org.teamapps.projector.components.trivial.multicurrency;

import com.fasterxml.jackson.databind.JsonNode;
import com.ibm.icu.util.ULocale;
import org.teamapps.projector.annotation.ClientObjectLibrary;
import org.teamapps.projector.component.field.AbstractField;
import org.teamapps.projector.component.field.DtoAbstractField;
import org.teamapps.projector.components.trivial.TrivialComponentsLibrary;
import org.teamapps.projector.components.trivial.dto.*;
import org.teamapps.projector.components.trivial.multicurrency.value.CurrencyUnit;
import org.teamapps.projector.components.trivial.multicurrency.value.CurrencyValue;
import org.teamapps.projector.event.ProjectorEvent;
import org.teamapps.projector.session.SessionContext;

import javax.annotation.Nonnull;
import java.math.BigDecimal;
import java.util.List;
import java.util.Locale;
import java.util.Optional;
import java.util.stream.Collectors;

@ClientObjectLibrary(value = TrivialComponentsLibrary.class)
public class CurrencyField extends AbstractField<CurrencyValue> implements DtoCurrencyFieldEventHandler {

	private final DtoCurrencyFieldClientObjectChannel clientObjectChannel = new DtoCurrencyFieldClientObjectChannel(getClientObjectChannel());

	public final ProjectorEvent<String> onTextInput = new ProjectorEvent<>(clientObjectChannel::toggleTextInputEvent);

	private ULocale locale = SessionContext.current().getULocale();

	private List<CurrencyUnit> currencies;

	/**
	 * If this is >= 0 it will overwrite the precision of the currencies.
	 */
	private int fixedPrecision = -1;

	private boolean alphabeticKeysQueryEnabled = true;

	/**
	 * If true, the currency indicator is on the left of the amount. If false, it is on the right of the amount.
	 */
	private boolean currencyBeforeAmount;

	/**
	 * Whether to show the currency symbol additionally to the currency code.
	 */
	private boolean currencySymbolsEnabled = true;

	public CurrencyField() {
		this(CurrencyUnit.getAllAvailableFromJdk());
	}

	public CurrencyField(List<CurrencyUnit> currencies) {
		this.currencies = currencies;
	}

	@Override
	public DtoAbstractField createConfig() {
		DtoCurrencyField field = new DtoCurrencyField();
		mapAbstractFieldAttributesToUiField(field);
		field.setCurrencyUnits(currencies.stream()
				.map(unit -> unit.toUiCurrencyUnit(locale.toLocale()))
				.collect(Collectors.toList()));
		field.setFixedPrecision(fixedPrecision);
		field.setShowCurrencyBeforeAmount(currencyBeforeAmount);
		field.setShowCurrencySymbol(currencySymbolsEnabled);
		field.setAlphabeticKeysQueryEnabled(alphabeticKeysQueryEnabled);
		field.setLocale(locale.toLanguageTag());
		return field;
	}

	public List<CurrencyUnit> getCurrencies() {
		return currencies;
	}

	public CurrencyField setCurrencies(List<CurrencyUnit> currencies) {
		this.currencies = currencies;
		clientObjectChannel.setCurrencyUnits(currencies != null ? currencies.stream()
				.map(unit -> unit.toUiCurrencyUnit(locale.toLocale()))
				.collect(Collectors.toList()) : null);
		return this;
	}

	@Override
	public Object convertServerValueToClientValue(CurrencyValue currencyValue) {
		return currencyValue != null ? currencyValue.toUiCurrencyValue(locale.toLocale()) : null;
	}

	@Override
	public CurrencyValue doConvertClientValueToServerValue(@Nonnull JsonNode value) {
		DtoCurrencyValueWrapper uiCurrencyValue = new DtoCurrencyValueWrapper(value);
		String uiAmount = uiCurrencyValue.getAmount();
		BigDecimal uxAmount = uiAmount != null ? new BigDecimal(uiAmount) : null;

		DtoCurrencyUnitWrapper uiCurrencyUnit = uiCurrencyValue.getCurrencyUnit();
		CurrencyUnit uxCurrencyUnit;
		if (uiCurrencyUnit != null) {
			uxCurrencyUnit = currencies.stream()
					.filter(cu -> cu.getCode().equals(uiCurrencyUnit.getCode()) && cu.getFractionDigits() == uiCurrencyUnit.getFractionDigits())
					.findFirst()
					.orElseGet(() -> CurrencyUnit.from(uiCurrencyUnit.getCode(), uiCurrencyUnit.getFractionDigits(), uiCurrencyUnit.getName(), uiCurrencyUnit.getSymbol()));
		} else {
			uxCurrencyUnit = null;
		}
		return new CurrencyValue(uxCurrencyUnit, uxAmount);
	}

	public Optional<CurrencyUnit> getCurrency() {
		CurrencyValue value = getValue();
		return value != null ? value.getCurrency() : Optional.empty();
	}

	public void setCurrency(CurrencyUnit currencyUnit) {
		CurrencyValue value = getValue();
		setValue(value != null ? value.withCurrencyUnit(currencyUnit) : new CurrencyValue(currencyUnit, null));
	}

	public Optional<BigDecimal> getAmount() {
		CurrencyValue value = getValue();
		return value != null ? value.getAmount() : Optional.empty();
	}

	public void setAmount(BigDecimal amount) {
		CurrencyValue value = getValue();
		setValue(value != null ? value.withAmount(amount) : new CurrencyValue(null, amount));
	}

	public int getFixedPrecision() {
		return fixedPrecision;
	}

	public boolean isCurrencyBeforeAmount() {
		return currencyBeforeAmount;
	}

	public CurrencyField setCurrencyBeforeAmount(boolean currencyBeforeAmount) {
		this.currencyBeforeAmount = currencyBeforeAmount;
		clientObjectChannel.setShowCurrencyBeforeAmount(currencyBeforeAmount);
		return this;
	}

	public boolean isCurrencySymbolsEnabled() {
		return currencySymbolsEnabled;
	}

	public CurrencyField setCurrencySymbolsEnabled(boolean currencySymbolsEnabled) {
		this.currencySymbolsEnabled = currencySymbolsEnabled;
		clientObjectChannel.setShowCurrencySymbol(currencySymbolsEnabled);
		return this;
	}

	public boolean isAlphabeticKeysQueryEnabled() {
		return alphabeticKeysQueryEnabled;
	}

	public void setFixedPrecision(int fixedPrecision) {
		this.fixedPrecision = fixedPrecision;
		clientObjectChannel.setFixedPrecision(fixedPrecision);
	}

	public void setAlphabeticKeysQueryEnabled(boolean alphabeticKeysQueryEnabled) {
		this.alphabeticKeysQueryEnabled = alphabeticKeysQueryEnabled;
		clientObjectChannel.setAlphabeticKeysQueryEnabled(alphabeticKeysQueryEnabled);
	}

	public Locale getLocale() {
		return locale.toLocale();
	}

	public ULocale getULocale() {
		return locale;
	}

	public void setLocale(Locale locale) {
		setULocale(ULocale.forLocale(locale));
	}

	public void setULocale(ULocale locale) {
		this.locale = locale;
		setCurrencies(currencies);
		clientObjectChannel.setLocale(locale.toLanguageTag());
	}

	@Override
	public void handleTextInput(String enteredString) {
		onTextInput.fire(enteredString);
	}
}

File: client/projector-tree-components/src/main/ts/CurrencyField.ts
/*-
 * ========================LICENSE_START=================================
 * TeamApps
 * ---
 * Copyright (C) 2014 - 2022 TeamApps.org
 * ---
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * =========================LICENSE_END==================================
 */
import {defaultListQueryFunctionFactory, isModifierKey, type QueryFunction} from "./trivial-components/TrivialCore";
import {TrivialUnitBox, type TrivialUnitBoxChangeEvent} from "./trivial-components/TrivialUnitBox";
import {
	AbstractField,
	BigDecimal,
	DebounceModes,
	deepEquals,
	executeAfterAttached,
	type FieldEditingMode,
	FieldEditingModes,
	ProjectorEvent
} from "projector-client-object-api";
import {
	type DtoCurrencyField,
	type DtoCurrencyField_TextInputEvent,
	type DtoCurrencyFieldCommandHandler,
	type DtoCurrencyFieldEventSource,
	type DtoCurrencyUnit,
	type DtoCurrencyValue
} from "./generated";
import {selectElementContents} from "./util";

export class CurrencyField extends AbstractField<DtoCurrencyField, DtoCurrencyValue> implements DtoCurrencyFieldEventSource, DtoCurrencyFieldCommandHandler {

	public readonly onTextInput: ProjectorEvent<DtoCurrencyField_TextInputEvent> = ProjectorEvent.createDebounced(250, DebounceModes.BOTH);

	private trivialUnitBox: TrivialUnitBox<DtoCurrencyUnit>;
	private queryFunction: QueryFunction<DtoCurrencyUnit>;

	protected initialize(config: DtoCurrencyField) {

		this.trivialUnitBox = new TrivialUnitBox<DtoCurrencyUnit>({
			numberFormatFunction: entry => this.getNumberFormat(entry),
			idFunction: entry => entry.code,
			unitDisplayPosition: config.showCurrencyBeforeAmount ? 'left' : 'right', // right or left
			entryRenderingFunction: entry => {
				entry = entry ?? {} as DtoCurrencyUnit;
				return `<div class="tr-template-currency-single-line-long">
				  <div class="content-wrapper tr-editor-area"> 
					<div class="symbol-and-code">${entry.code != null ? `<span class="currency-code">${entry.code || ''}</span>` : ''} ${entry.symbol != null ? `<span class="currency-symbol">${entry.symbol || ''}</span>` : ''}</div>
					<div class="currency-name">${entry.name || ''}</div>
				  </div>
				</div>`;
			},
			selectedEntryRenderingFunction: (entry) => {
				if (entry == null) {
					return `<div class="tr-template-currency-single-line-short">-</div>`
				} else if (this.config.showCurrencySymbol && entry.symbol) {
					return `<div class="tr-template-currency-single-line-short">${entry.symbol} (${entry.code})</div>`;
				} else {
					return `<div class="tr-template-currency-single-line-short">${entry.code}</div>`;
				}
			},
			queryOnNonNumberCharacters: config.alphabeticKeysQueryEnabled,
			editingMode: this.convertToTrivialComponentsEditingMode(config.editingMode),
			queryFunction: (queryString) => this.queryFunction(queryString)
		});
		this.trivialUnitBox.getMainDomElement().classList.add("DtoCurrencyField", "default-min-field-width");
		this.trivialUnitBox.onChange.addListener((value: TrivialUnitBoxChangeEvent<any>) => {
			this.commit();
		});
		this.trivialUnitBox.getEditor().addEventListener('keyup', (e: KeyboardEvent) => {
			if (e.key !== "Enter"
				&& e.key !== "Tab"
				&& !isModifierKey(e)) {
				this.onTextInput.fire({
					enteredString: (this.trivialUnitBox.getEditor() as HTMLInputElement).value
				});
			}
		});

		this.setCurrencyUnits(config.currencyUnits);

		this.trivialUnitBox.getMainDomElement().classList.add("field-border", "field-border-glow", "field-background");
		this.trivialUnitBox.getMainDomElement().querySelector<HTMLElement>(":scope .tr-editor").classList.add("field-background");
		this.trivialUnitBox.getMainDomElement().querySelector<HTMLElement>(":scope .tr-unitbox-selected-entry-and-trigger-wrapper").classList.add("field-border");
		this.trivialUnitBox.onFocus.addListener(() => this.getMainElement().classList.add("focus"));
		this.trivialUnitBox.onBlur.addListener(() => this.getMainElement().classList.remove("focus"));
	}

	private getNumberFormat(entry: DtoCurrencyUnit) {
		if (entry == null) {
			let fractionDigits = this.config.fixedPrecision >= 0 ? this.config.fixedPrecision : 2;
			return new Intl.NumberFormat(this.config.locale, {
				useGrouping: true,
				minimumFractionDigits: fractionDigits,
				maximumFractionDigits: fractionDigits
			});
		} else {
			let fractionDigits = this.config.fixedPrecision >= 0 ? this.config.fixedPrecision : entry.fractionDigits >= 0 ? entry.fractionDigits : 4;
			return new Intl.NumberFormat(this.config.locale, {
				minimumFractionDigits: fractionDigits,
				maximumFractionDigits: fractionDigits,
				useGrouping: true
			})
		}
	}

	isValidData(v: DtoCurrencyValue): boolean {
		return v == null || typeof v === "object";
	}

	private convertToTrivialComponentsEditingMode(editingMode: FieldEditingMode) {
		return editingMode === FieldEditingModes.READONLY ? 'readonly' : editingMode === FieldEditingModes.DISABLED ? 'disabled' : 'editable';
	}

	public getMainInnerDomElement(): HTMLElement {
		return this.trivialUnitBox.getMainDomElement() as HTMLElement;
	}

	protected displayCommittedValue(): void {
		let DtoValue = this.getCommittedValue();
		if (DtoValue) {
			this.trivialUnitBox.setSelectedEntry(DtoValue?.currencyUnit);
			this.trivialUnitBox.setAmount(BigDecimal.of(DtoValue?.amount));
		} else {
			this.trivialUnitBox.setAmount(null);
			this.trivialUnitBox.setSelectedEntry(null);
		}
	}

	@executeAfterAttached()
	focus(): void {
		this.trivialUnitBox.focus();
		selectElementContents(this.trivialUnitBox.getMainDomElement().querySelector<HTMLElement>(":scope .tr-editor"));
	}

	destroy(): void {
		super.destroy();
		this.trivialUnitBox.destroy();
	}

	getTransientValue(): DtoCurrencyValue {
		return {
			currencyUnit: this.trivialUnitBox.getSelectedEntry() && this.trivialUnitBox.getSelectedEntry(),
			amount: this.trivialUnitBox.getAmount()?.value
		};
	}

	protected onEditingModeChanged(editingMode: FieldEditingMode): void {
		this.getMainElement().classList.remove(...Object.values(FieldEditingModes));
		this.getMainElement().classList.add(editingMode);
		this.trivialUnitBox.setEditingMode(this.convertToTrivialComponentsEditingMode(editingMode));
	}

	public getReadOnlyHtml(value: DtoCurrencyValue, availableWidth: number): string {
		let content: string;
		if (value != null) {
			const currency = value.currencyUnit;
			let displayedCurrency: string;
			if (currency != null) {
				displayedCurrency = this.config.showCurrencySymbol ? `${currency?.symbol} (${currency?.code})` : currency?.code;
			} else {
				displayedCurrency = null;
			}
			let amount = BigDecimal.of(value.amount);
			let formattedAmount = amount?.format(this.getNumberFormat(currency));
			if (this.config.showCurrencyBeforeAmount) {
				content = [displayedCurrency, formattedAmount].filter(x => x != null).join(' ');
			} else {
				content = [formattedAmount, displayedCurrency].filter(x => x != null).join(' ');
			}
		} else {
			content = "";
		}
		return `<div class="static-readonly-DtoCurrencyField">${content}</div>`;
	}

	getDefaultValue() {
		return {
			currencyUnit: null, amount: null
		};
	}

	setCurrencyUnits(currencyUnits: DtoCurrencyUnit[]): void {
		this.config.currencyUnits = currencyUnits;
		this.queryFunction = defaultListQueryFunctionFactory(currencyUnits, ["code", "name", "symbol"], {matchingMode: "contains", ignoreCase: true});
	}

	setShowCurrencyBeforeAmount(showCurrencyBeforeAmount: boolean): void {
		this.trivialUnitBox.setUnitDisplayPosition(showCurrencyBeforeAmount ? 'left' : "right")
	}

	setShowCurrencySymbol(showCurrencySymbol: boolean): void {
		this.config.showCurrencySymbol = showCurrencySymbol;
		this.trivialUnitBox.setSelectedEntry(this.trivialUnitBox.getSelectedEntry());
	}

	public valuesChanged(v1: DtoCurrencyValue, v2: DtoCurrencyValue): boolean {
		return !deepEquals(v1, v2);
	}

	setLocale(locale: string): void {
		this.config.locale = locale;
		this.trivialUnitBox.setSelectedEntry(this.trivialUnitBox.getSelectedEntry()); // update format
	}

	setFixedPrecision(fixedPrecision: number): void {
		this.config.fixedPrecision = fixedPrecision;
		this.trivialUnitBox.setSelectedEntry(this.trivialUnitBox.getSelectedEntry()); // update format
	}

	setAlphabeticKeysQueryEnabled(alphabeticKeysQueryEnabled: boolean) {
		this.config.alphabeticKeysQueryEnabled = alphabeticKeysQueryEnabled;
		this.trivialUnitBox.setQueryOnNonNumberCharacters(alphabeticKeysQueryEnabled);
	}
}


